api/envoy/extensions/filters/http/rate_limit_quota/v3/rate_limit_quota.proto

```diff
@@ -0,0 +1,160 @@
+syntax = "proto3";
+
+package envoy.extensions.filters.http.rate_limit_quota.v3;
+
+import "envoy/config/core/v3/base.proto";
+import "envoy/config/core/v3/config_source.proto";
+import "envoy/config/core/v3/extension.proto";
+import "envoy/type/v3/ratelimit_strategy.proto";
+
+import "google/protobuf/duration.proto";
+import "google/rpc/status.proto";
+
+import "xds/annotations/v3/status.proto";
+import "xds/type/matcher/v3/matcher.proto";
+
+import "udpa/annotations/status.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.filters.http.rate_limit_quota.v3";
+option java_outer_classname = "RateLimitQuotaProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/rate_limit_quota/v3;rate_limit_quotav3";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+option (xds.annotations.v3.file_status).work_in_progress = true;
+
+// [#not-implemented-hide:]
+// [#protodoc-title: Rate Limit Quota]
+// Rate limit :ref:`configuration overview <config_http_filters_rate_limit_quota>`.
+// [#extension: envoy.filters.http.rate_limit_quota]
+
+message RateLimitQuota {
+  // [#comment:TODO(sergiitk): preview mode matchers]
+
+  // Configuration for the gRPC service that hosts an external rate limit
+  // provider. If not specified, any calls to the rate limit service will
+  // immediately return success.
+  config.core.v3.ConfigSource rlqs_server = 1;
+
+  // The application domain to use when calling the service.
+  string domain = 2;
+
+  // Unified matchers.
+  xds.type.matcher.v3.Matcher bucket_matchers = 3;
+}
+
+// [#next-free-field: 7]
+message RateLimitQuotaBucketSettings {
+  // Configures the behavior when the first request has been matched to
+  // the bucket, and the data plane is waiting for the RLQS to return the
+  // first quota assignment.
+  message NoAssignmentBehavior {
```

Comment:
I like this idea. Will do.

Reply:
Should this be somehow pre-warmed? E.g. hold filter loading until RLQS responds?

Reply:
Just to clarify it regarding the `blocking behavior` in Envoy:

We just don't need the `RateLimitStrategyQueue` to achieve the blocking behavior, instead it is done implicitly in Envoy based on timer: The filter for each request will stop the filter chain iteration while waiting for the response (which is like queuing), and resume the filter chain once we get the response or abort it times out (HTTP connection manager will keep track of the active requests and delete the timed-out ones).

Reply:
Re: my question. I was specifically thinking of listener warming. When this filter config is loaded, the filter can indicate it's not ready and perform some async work. E.g. it can fetch assignments. However, I think it's probably not always a good idea since it'll consume quota without any requests.

Reply:
Re: predefined quota for the new requests. Addressed in https://github.com/envoyproxy/envoy/pull/19793/commits/0e3eaf3b634e52f6966a7e32ef202b2d6fbc42e6.